### PR TITLE
Fix #257581, Fix #64636  - Changes to Measure properties are not propagated between score and existing linked parts

### PR DESCRIPTION
**Pull request by njvdberg**
Resolves: https://musescore.org/en/node/257581
Resolves: https://musescore.org/en/node/64636

When reading a mscx file containing parts, link all measures of the part to the corresponding measure in the MasterScore.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: libmscore/read301.cpp [hunk 1]
```diff
@@ -289,6 +291,27 @@ bool Score::read(XmlReader& e)
...
+void Score::linkMeasures(Score* score)
+      {
+      MeasureBase *mbMaster = score->first();
+      for (MeasureBase* mb = first(); mb; mb = mb->next()) {
+            if (!mb->isMeasure())
+                  continue;
+            while (mbMaster && !mbMaster->isMeasure())
+                  mbMaster = mbMaster->next();
+            if (!mbMaster) {
+                  qDebug("Measuses in MasterScoare and Score are not in sync.");
```
Comment (Jojo-Schmitz) [inline libmscore/read301.cpp:307]: typo

Comment (njvdberg) [reply libmscore/read301.cpp]: Oops

**Comment by Jojo-Schmitz** [comment]
I wonder whether this is needed in read206.cpp too?

**Comment by njvdberg** [comment]
That part of the code looks very similar in both read206.cpp and read301.cpp indeed. Will check with a MuseScore 2.x file.

**Comment by njvdberg** [comment]
> I wonder whether this is needed in read206.cpp too? (certainly not in read114.cpp though)
Just checked, same change is **not** required in <code>read206.cpp</code>. As a matter of fact, this issue is a regression since 3.0 (educated guess, not checked). In the 2.06 version of the file, the file number was included in the <code>Measure</code> tag as an attribute <code>number</code>. This attribute is removed in file version 3.01, introducing the issue.

**Comment by Jojo-Schmitz** [comment]
OK, thanks for checking!

File: libmscore/read301.cpp [hunk 2]
```diff
@@ -289,6 +291,27 @@ bool Score::read(XmlReader& e)
       return true;
       }
 
+//---------------------------------------------------------
+// linkMeasures
+//---------------------------------------------------------
+
+void Score::linkMeasures(Score* score)
+      {
+      MeasureBase *mbMaster = score->first();
```
Comment (vpereverzev) [inline libmscore/read301.cpp:300]: I think MeasureBaseList is a good candidate for an implementation of this method

So here you can leave something like that:
`if (!score || _measure.size() == 0)
      return;

_measures->link(score->measures());
`

Also I'd suggest to add a very simple method to MeasureBaseList - "bool hasNext() const;"
It'll slightly improve your loops.

**Comment by Jojo-Schmitz** [comment]
rebase needed

**Comment by njvdberg** [comment]
Rebase done

**Review by Jojo-Schmitz — APPROVED** [review]

**Pull request merged**